### PR TITLE
Add advance settings to fine tune DRS imbalance calculation

### PR DESCRIPTION
**Pull request by vishesh92**
### Description
Doc PR: https://github.com/apache/cloudstack-documentation/pull/374

This PR addresses two issues:
1. Allow DRS to work as expected in clusters with hosts of different capacities - We do this by using ratio of free/total metric. This is controlled using a new setting `drs.metric.use.ratio`.
2. Imbalance can be greater than 1. For condensed algorithm, migrations don't take place when this happens. - This happens when the distribution of metrics is skewed. We fix this by filtering out values where ratio is less than `drs.imbalance.condensed.skip.threshold` if `drs.metric.type` is `free` and greater than `drs.imbalance.condensed.skip.threshold` if `drs.metric.type` is `used` to remove the skew and check if we need drs or not.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@blueorangutan test

File: api/src/main/java/org/apache/cloudstack/cluster/ClusterDrsAlgorithm.java [hunk 1]
```diff
@@ -87,36 +90,65 @@ Ternary<Double, Double, Double> getMetrics(long clusterId, VirtualMachine vm, Se
...
+                vmMetric = (long) serviceOffering.getCpu() * serviceOffering.getSpeed();
+                break;
+            case "memory":
+                hostMetricsMap = hostMemoryMap;
+                vmMetric = serviceOffering.getRamSize() * 1024L * 1024L;
+                break;
+            default:
+                throw new ConfigurationException(
+                        String.format("Invalid metric: %s for cluster: %d", metric, destHost.getClusterId()));
+        }
```
Comment (DaanHoogland) [inline api/src/main/java/org/apache/cloudstack/cluster/ClusterDrsAlgorithm.java:118]: could be a `setMetricsMapAndType()`

File: api/src/main/java/org/apache/cloudstack/cluster/ClusterDrsAlgorithm.java [hunk 2]
```diff
@@ -87,36 +90,65 @@ Ternary<Double, Double, Double> getMetrics(long clusterId, VirtualMachine vm, Se
...
+                free -= vmMetric;
             } else if (hostId.equals(vm.getHostId())) {
-                postCpuList.add((double)(cpu + vmCpu)/hostCpuFreeMap.get(hostId).second());
-                postMemoryList.add((double)(memory + vmRam)/hostMemoryFreeMap.get(hostId).second());
-            } else {
-                postCpuList.add((double)cpu/hostCpuFreeMap.get(hostId).second());
-                postMemoryList.add((double)memory/hostMemoryFreeMap.get(hostId).second());
+                used -= vmMetric;
+                free += vmMetric;
+            }
```
Comment (DaanHoogland) [inline api/src/main/java/org/apache/cloudstack/cluster/ClusterDrsAlgorithm.java:134]: `getHostBalanceData()

File: api/src/main/java/org/apache/cloudstack/cluster/ClusterDrsAlgorithm.java [hunk 3]
```diff
@@ -164,4 +213,31 @@ default Double getClusterStandardDeviation(List<Double> metricList, Double mean)
...
+                    }
+                case "used":
+                    if (skipThreshold != null && used > skipThreshold * actualTotal) continue;
+                    if (useRatio) {
+                        list.add((double) used / actualTotal);
+                    } else {
+                        list.add((double) used);
+                    }
+            }
+        }
```
Comment (DaanHoogland) [inline api/src/main/java/org/apache/cloudstack/cluster/ClusterDrsAlgorithm.java:240]: `addHostValueToImbalanceList()` but with skip flag

File: plugins/drs/cluster/condensed/src/main/java/org/apache/cloudstack/cluster/Condensed.java [hunk 4]
```diff
@@ -21,78 +21,61 @@
...
 import javax.naming.ConfigurationException;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
 
 import static org.apache.cloudstack.cluster.ClusterDrsService.ClusterDrsImbalanceThreshold;
-import static org.apache.cloudstack.cluster.ClusterDrsService.ClusterDrsMetric;
+import static org.apache.cloudstack.cluster.ClusterDrsService.ClusterDrsMetricType;
 
 public class Condensed extends AdapterBase implements ClusterDrsAlgorithm {
```
Comment (harikrishna-patnala) [inline plugins/drs/cluster/condensed/src/main/java/org/apache/cloudstack/cluster/Condensed.java:37]: @vishesh92 you will need to extend Configurable interface and add "ClusterDrsImbalanceSkipThreshold" in getConfigKeys() method to save this new config in the database.

**Comment by kiranchavala** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan package

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan ui

**Comment by kiranchavala** [comment]
@blueorangutan test

**Review by kiranchavala — APPROVED** [review]
Tested the new Advanced DRS related settings manually and also the UI improvements


drs.metric.type
drs.metric.use.ratio
drs.imbalance.condensed.skip.threshold

Scenarios

Drs.imbalance= 0.9 and drs.alogirithm=condensed >>  The vms got migrated on a single host

Drs.imbalance= 0.4 and drs.alogirthm=balanced >>  The vms got distributed across the hosts


Screenshots

resize

drsvalues1

drsvalues2

Logs


Condensed


```
2024-02-08 05:35:54,311 DEBUG [o.a.c.c.Balanced] (qtp1491755116-22:ctx-ecfb388b ctx-9739cec5) (logid:8253e2f5) Cluster 2 needs DRS. Imbalance: 0.7158188976374373 Threshold: 0.5 Algorithm: balanced DRS metric: memory Metric Type: used Use ratio: false

2024-02-08 05:35:54,405 DEBUG [o.a.c.c.Balanced] (qtp1491755116-22:ctx-ecfb388b ctx-9739cec5) (logid:8253e2f5) Cluster 2 pre-imbalance: 0.7158188976374373 post-imbalance: 0.7820295697311479 Algorithm: balanced VM: f659a6f7-2915-4a18-9217-11cf8a03dce1 srcHost: 4 destHost: a16c16d2-a2bb-4dbf-93fb-838568f3a2fd

2024-02-08 05:35:54,413 DEBUG [o.a.c.c.Balanced] (qtp1491755116-22:ctx-ecfb388b ctx-9739cec5) (logid:8253e2f5) Cluster 2 pre-imbalance: 0.7158188976374373 post-imbalance: 0.34015067152490375 Algorithm: balanced VM: f659a6f7-2915-4a18-9217-11cf8a03dce1 srcHost: 4 destHost: b1f62d2b-fee5-4fab-b3c1-40ab3ad68a45

2024-02-08 05:35:54,413 DEBUG [o.a.c.c.ClusterDrsServiceImpl] (qtp1491755116-22:ctx-ecfb388b ctx-9739cec5) (logid:8253e2f5) Plan for VM e1096698-c0f3-4c75-8c9a-fc3526c938dd to migrate from host 722009bc-e65f-465c-a55f-1e26b70b93ae to host b1f62d2b-fee5-4fab-b3c1-40ab3ad68a45

2024-02-08 05:35:54,417 DEBUG [o.a.c.c.Balanced] (qtp1491755116-22:ctx-ecfb388b ctx-9739cec5) (logid:8253e2f5) Cluster 2 does not need DRS. Imbalance: 0.34015067152490375 Threshold: 0.5 Algorithm: balanced DRS metric: memory Metric Type: used Use ratio: false
```


Balanced
```
2024-02-08 05:49:16,865 DEBUG [o.a.c.c.Condensed] (qtp1491755116-22:ctx-dbbaeaf5 ctx-975fb795) (logid:d6ccccf6) Cluster 2 needs DRS. Imbalance: 0.34015067152490375 Threshold: 0.8999999761581421 Algorithm: condensed DRS metric: memory Metric Type: used Use ratio: false SkipThreshold: 0.95

2024-02-08 05:49:17,055 DEBUG [o.a.c.c.Condensed] (qtp1491755116-22:ctx-dbbaeaf5 ctx-975fb795) (logid:d6ccccf6) Cluster 2 pre-imbalance: 0.7820295697311479 post-imbalance: 1.0523488093445659 Algorithm: condensed VM: f659a6f7-2915-4a18-9217-11cf8a03dce1 srcHost: 4 destHost: a16c16d2-a2bb-4dbf-93fb-838568f3a2fd

2024-02-08 05:49:17,060 DEBUG [o.a.c.c.Condensed] (qtp1491755116-22:ctx-dbbaeaf5 ctx-975fb795) (logid:d6ccccf6) Cluster 2 pre-imbalance: 0.7820295697311479 post-imbalance: 0.6428243465332251 Algorithm: condensed VM: f659a6f7-2915-4a18-9217-11cf8a03dce1 srcHost: 4 destHost: b1f62d2b-fee5-4fab-b3c1-40ab3ad68a45

2024-02-08 05:49:16,965 DEBUG [o.a.c.c.ClusterDrsServiceImpl] (qtp1491755116-22:ctx-dbbaeaf5 ctx-975fb795) (logid:d6ccccf6) Plan for VM e1096698-c0f3-4c75-8c9a-fc3526c938dd to migrate from host b1f62d2b-fee5-4fab-b3c1-40ab3ad68a45 to host a16c16d2-a2bb-4dbf-93fb-838568f3a2fd

2024-02-08 05:49:17,060 DEBUG [o.a.c.c.ClusterDrsServiceImpl] (qtp1491755116-22:ctx-dbbaeaf5 ctx-975fb795) (logid:d6ccccf6) Plan for VM 0381cd82-f9dd-417d-836f-99dd300720e7 to migrate from host 722009bc-e65f-465c-a55f-1e26b70b93ae to host a16c16d2-a2bb-4dbf-93fb-838568f3a2fd

2024-02-08 05:49:17,064 DEBUG [o.a.c.c.Condensed] (qtp1491755116-22:ctx-dbbaeaf5 ctx-975fb795) (logid:d6ccccf6) Cluster 2 does not need DRS. Imbalance: 1.0523488093445659 Threshold: 0.8999999761581421 Algorithm: condensed DRS metric: memory Metric Type: used Use ratio: false SkipThreshold: 0.95
```

**Comment by rohityadavcloud** [comment]
Is this good to merge @kiranchavala @vishesh92 ?

**Comment by rohityadavcloud** [comment]
LGTM, thanks for the fixes and testing @vishesh92 @kiranchavala

**Pull request merged**